### PR TITLE
Multi-modal Cluttered Scene Analysis in Knowledge Intensive Scenarios (GSOC 2017)

### PR DESCRIPTION
**Pull request by anindex**
**Pull request summaries**
Implemented AEs:
- OverSegmentationAnnotator
- RotationalSymmetryAnnotator
- RotationalSymmetrySegmentation
- BilateralSymmetryAnnotator
- BilateralSymmetrySegmentation
- PointNormalNaNFilter
- SegmentationResultDisplayer

Pipelines:
- full_segmentation (combining rotational and bilateral symmetry segmentation on low noise clouds)
- object_segmentation (bilateral symmetry segmentation on raw clouds)

Datasets:
- 4 low noise clouds in author's dataset: www.umiacs.umd.edu/%7Eaecins/projects/symseg/dataset/CTD.tar.gz
- 2 raw clouds in TUW dataset: http://www.acin.tuwien.ac.at/?id=289
 
Unit-test:
- Two unit-test on low noise cloud dataset (cloud2 and cloud3) with ground-truths.
- Criteria to pass: successfully segment at least 75% of total objects in the scene, each segments has at least 70% point similarity with ground-truth table.
- Run `catkin_make run_tests_robosherlock` to run unit-testing

Documentation:
- In-code descriptions
- Overview and usage in this [link](https://gist.github.com/anindex/ee4b32e73973ab24d3d7a17ab6486b9d)

### CODE REVIEW

**Comment by anindex** [comment]
The runtime unit-test in travis check is time-out, maybe low-performance virtual machine causes this situation. I have tested the unit-tests at my own computer with runtime 2-3s per unit-test.
Should I exclude unit-test out of travis? Because these tests need computation power.

**Comment by bbferka** [comment]
I would maybe put it in the Normal Estimator

**Comment by anindex** [comment]
I met some weird behaviors of rs::conversion::from. When I tried to get cloud non NaN and the indices:
`rs::ReferenceClusterPoints rcp = rs::create<rs::ReferenceClusterPoints>(tcas);
 cas.get(VIEW_CLOUD_NON_NAN, rcp);
 rs::conversion::from(rcp.cloud(), *cloud_ptr);
 rs::conversion::from(rcp.indices.get(), mapping_to_original);`

The pipeline says that: 
`DrawingAnnotator.cpp(79)[process] Exception in SegmentationResultDisplayer: 
Exception     : FeatureNotAppropriateException

Error number  : 5025
Recoverable   : Yes
Error         : Feature 'indices' not appropriate for type 'rs.pcl.PointCloud'
   While      : Getting value of a feature
`

mapping_to_original is type pcl::PointIndices, indices in pcl_type is also rs.pcl.PointIndices

**Comment by bbferka** [comment]
use getFS instead; this is a weird bug in RoboSherlock;

File: src/utils/src/RSAnalysisEngine.cpp [hunk 1]
```diff
@@ -41,11 +41,6 @@ RSAnalysisEngine::~RSAnalysisEngine()
     delete engine;
     engine = NULL;
   }
-  if(rspm)
-  {
-    delete rspm;
-    rspm = NULL;
-  }
 }
```
Comment (bbferka) [inline src/utils/src/RSAnalysisEngine.cpp]: @anindex  did this have anything to do with why your unit test were failing?

Comment (anindex) [reply src/utils/src/RSAnalysisEngine.cpp]: I just remove the redundant code, cause it is a duplication. And it also has no effect on the failing, so if it causes some problem I will put it back

Comment (bbferka) [reply src/utils/src/RSAnalysisEngine.cpp]: oh, sorry, I did not notice that a few lines above it is already deleted :)

**Comment by bbferka** [comment]
@anindex  if you fix such that travis builds and you move the graph implementations to utils I would merge this PR

**Comment by anindex** [comment]
I will do that tomorrow.

**Pull request merged**